ext/digest/crc/crc_ext.c

```diff
@@ -0,0 +1,431 @@
+#define table_initializeR64(bits, n)    (n)
+#define bitwise_shiftR64(n)             ((n) >> 1)
+#define bitwise_is_carryoverR64(n)      ((n) & 1)
+
+#define UINT2NUM8(n)                    UINT2NUM(n)
+#define UINT2NUM16(n)                   UINT2NUM(n)
+#define UINT2NUM32(n)                   UINT2NUM(n)
+#define UINT2NUM64(n)                   ULL2NUM(n)
+#define NUM2UINT8(v)                    (NUM2UINT(v) & 0xff)
+#define NUM2UINT16(v)                   (NUM2UINT(v) & 0xffff)
+#define NUM2UINT32(v)                   (NUM2UINT(v) & 0xffffffff)
+#define NUM2UINT64(v)                   (NUM2ULL(v) & 0xffffffffffffffff)
+
+typedef uint8_t UINT8;
+typedef uint16_t UINT16;
+typedef uint32_t UINT32;
+typedef uint64_t UINT64;
+
+static uint64_t
+update_crc(const struct crc_model *model, uint64_t crc, const void *buf, size_t len)
+{
+#if SLICE_SIZE == 8
+#define UPDATE_CRC_SLICING_CORE(V) \
+	table[7][p[0] ^ V(slice0)(state)] ^ \
+	table[6][p[1] ^ V(slice1)(state)] ^ \
+	table[5][p[2] ^ V(slice2)(state)] ^ \
+	table[4][p[3] ^ V(slice3)(state)] ^ \
+	table[3][p[4] ^ V(slice4)(state)] ^ \
+	table[2][p[5] ^ V(slice5)(state)] ^ \
+	table[1][p[6] ^ V(slice6)(state)] ^ \
+	table[0][p[7] ^ V(slice7)(state)]
+#else
+#define UPDATE_CRC_SLICING_CORE(V) \
+	table[15][p[ 0] ^ V(slice0)(state)] ^ \
+	table[14][p[ 1] ^ V(slice1)(state)] ^ \
+	table[13][p[ 2] ^ V(slice2)(state)] ^ \
+	table[12][p[ 3] ^ V(slice3)(state)] ^ \
+	table[11][p[ 4] ^ V(slice4)(state)] ^ \
+	table[10][p[ 5] ^ V(slice5)(state)] ^ \
+	table[ 9][p[ 6] ^ V(slice6)(state)] ^ \
+	table[ 8][p[ 7] ^ V(slice7)(state)] ^ \
+	table[ 7][p[ 8]                   ] ^ \
+	table[ 6][p[ 9]                   ] ^ \
+	table[ 5][p[10]                   ] ^ \
+	table[ 4][p[11]                   ] ^ \
+	table[ 3][p[12]                   ] ^ \
+	table[ 2][p[13]                   ] ^ \
+	table[ 1][p[14]                   ] ^ \
+	table[ 0][p[15]                   ]
+#endif
+
+#define UPDATE_CRC_CORE_DECL(V) \
+	crc_int_t state = V(enter_update)(model->bitwidth, (crc_int_t)crc); \
+	\
+	for (; len >= SLICE_SIZE; len -= SLICE_SIZE, p += SLICE_SIZE) { \
+		state = UPDATE_CRC_SLICING_CORE(V); \
+	} \
+	\
+	for (; len > 0; len--, p++) { \
+		state = table[0][p[0] ^ V(slice0)(state)] ^ \
+			V(shift)(state); \
+	} \
+	\
+	return (uint64_t)(V(leave_update)(model->bitwidth, state));
+
+#define UPDATE_CRC_DECL(V) \
+	typedef V(UINT) crc_int_t; \
+	typedef crc_int_t crc_table_t[256]; \
+	const crc_table_t *table = (const crc_table_t *)model->table; \
+	const uint8_t *p = (const uint8_t *)buf; \
+	\
+	DISPATCH_BY_INPUT_DIRECTION(model->reflect_input, V, UPDATE_CRC_CORE_DECL);
+
+	DISPATCH_BY_BITWIDTH(model->bitwidth, UPDATE_CRC_DECL);
+}
+
+static void
+initialize_crc_table(struct crc_model *model)
+{
+#define BUILD_TABLE_CORE_DECL(V) \
+	const crc_int_t modified_poly = V(adapt)(model->bitwidth, model->polynomial); \
+	\
+	for (int i = 0; i < 256; i++) { \
```

Comment:
Also an old school optimization trick is to always define your loop variables above/*outside* of your loop. This prevents dumb compilers for re-allocating space on the stack for nested for loop index variables.